### PR TITLE
Issue 36. Users can activate their account using their unique voter t…

### PR DESCRIPTION
**Pull request by amymariewall**
…oken and set a password.

### CODE REVIEW

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
 
     # See https://administrate-docs.herokuapp.com/customizing_controller_actions
     # for more information
+
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:19]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
...
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:20]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb [hunk 3]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
...
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:21]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb [hunk 4]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
...
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:22]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb [hunk 5]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
...
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:23]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb [hunk 6]
```diff
@@ -15,5 +15,12 @@ class UsersController < Admin::ApplicationController
...
+
```
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra blank line detected.

Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra empty line detected at class body end.

File: app/models/user.rb [hunk 7]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
   validates :first_name, presence: true
   validates :last_name, presence: true
 
+# The following lines were created using this as a resource: http://blog.bigbinary.com/2016/03/23/has-secure-  token-to-generate-unique-random-token-in-rails-5.html. I opted NOT to use the has_secure_token method because it doesn't appear that you can make tokens less than 24 characters, which would be difficult to use since might still want to manually enter them on paper forms/ballots.
```
Comment (sumeetjain) [inline app/models/user.rb]: Incorrect indentation detected (column 0 instead of 2).

Comment (sumeetjain) [inline app/models/user.rb]: Line is too long. [390/80]

File: app/models/user.rb [hunk 8]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
...
+
+  before_create :set_voter_token
```
Comment (sumeetjain) [inline app/models/user.rb:16]: Inconsistent indentation detected.

File: app/models/user.rb [hunk 9]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
...
+
+  private
+
+  def set_voter_token
```
Comment (sumeetjain) [inline app/models/user.rb:20]: Inconsistent indentation detected.

File: app/models/user.rb [hunk 10]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
...
+    self.voter_token = generate_token
+  end
+
+  def generate_token
```
Comment (sumeetjain) [inline app/models/user.rb:24]: Inconsistent indentation detected.

File: app/models/user.rb [hunk 11]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
...
+    loop do
+      token = SecureRandom.hex(4)
+      break token unless User.where(voter_token: token).exists?
+    end
+  end
+
+
```
Comment (sumeetjain) [inline app/models/user.rb:31]: Extra blank line detected.

File: app/models/user.rb [hunk 12]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
...
+
```
Comment (sumeetjain) [inline app/models/user.rb:32]: Extra blank line detected.

Comment (sumeetjain) [inline app/models/user.rb:32]: Extra empty line detected at class body end.

File: app/controllers/admin/users_controller.rb (hunk 5 above)
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:23]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb (hunk 6 above)
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra blank line detected.

Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra empty line detected at class body end.

File: app/models/user.rb [hunk 13]
```diff
@@ -13,4 +13,23 @@ class User < ActiveRecord::Base
   validates :first_name, presence: true
   validates :last_name, presence: true
 
+# The following lines were created using this as a resource: http://blog.bigbinary.com/2016/03/23/has-secure-  token-to-generate-unique-random-token-in-rails-5.html. I opted NOT to use the has_secure_token method/gem because it doesn't appear that you can make tokens less than 24 characters, which would be difficult to use since might still want to manually enter them on paper forms/ballots.
```
Comment (sumeetjain) [inline app/models/user.rb:14]: Line is too long. [394/80]

File: app/controllers/admin/users_controller.rb (hunk 5 above)
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:23]: Extra blank line detected.

File: app/controllers/admin/users_controller.rb (hunk 6 above)
Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra blank line detected.

Comment (sumeetjain) [inline app/controllers/admin/users_controller.rb:24]: Extra empty line detected at class body end.

File: config/routes.rb [hunk 14]
```diff
@@ -14,9 +14,15 @@
   get "users/home" => 'users#home'
 
 
-  devise_for :users
+  #devise_for :users
```
Comment (sumeetjain) [inline config/routes.rb:28]: Missing space after #.

File: config/routes.rb [hunk 15]
```diff
@@ -14,9 +14,15 @@
...
 
-  "home#index"
+  #{}"home#index"
```
Comment (sumeetjain) [inline config/routes.rb:30]: Missing space after #.

File: config/routes.rb [hunk 16]
```diff
@@ -14,9 +14,15 @@
...
+
+  devise_for :users, :skip => [:registrations]
```
Comment (sumeetjain) [inline config/routes.rb]: Use the new Ruby 1.9 hash syntax.

Comment (sumeetjain) [inline config/routes.rb]: Trailing whitespace detected.

Comment (sumeetjain) [inline config/routes.rb]: Line is too long. [88/80]

File: config/routes.rb [hunk 17]
```diff
@@ -14,9 +14,15 @@
...
 
 
-  devise_for :users
+  #devise_for :users
 
-  "home#index"
+  #{}"home#index"
+
+  devise_for :users, :skip => [:registrations]                                          
+    as :user do
```
Comment (sumeetjain) [inline config/routes.rb]: Inconsistent indentation detected.

File: config/routes.rb [hunk 18]
```diff
@@ -14,9 +14,15 @@
...
+      get 'users/edit' => 'devise/registrations#edit', :as => 'edit_user_registration'
```
Comment (sumeetjain) [inline config/routes.rb]: Line is too long. [90/80]

Comment (sumeetjain) [inline config/routes.rb]: Trailing whitespace detected.

**Comment by amymariewall** [comment]
I am currently looking for help using Devise/Administrate to create new Users that have a voter_token but not a password.
- [ ] Using Administrate, admins should be able to create new Users
- [x] A voter_token gets automatically generated and assigned when a new User is created (this is done using SecureRandom and methods in the User model)
- [x] Users then can access a page where they can edit their account
- [ ] This page should confirm them using their voter_token and allow them to set a password for their account.

I added a method to our routes that skips registration for users altogether, and I am not sure how to make it so admins can create new users but non-admins cannot.

**Comment by amymariewall** [comment]
Currently getting an error trying to render and use the devise edit password page. I could use some help figuring out my next step.
Error

**Pull request merged**